torch/distributed/distributed_c10d.py

```diff
@@ -1064,6 +1064,159 @@ def recv(tensor, src=None, group=None, tag=0):
         return src
 
 
+Tensordtype2ncclDataType_t={
+    # dtype              # corresponding to ncclDataType_t from `nccl.h`
+    "torch.int8"    : 0, # ncclInt8       = 0
+    "torch.uint8"   : 1, # ncclUint8      = 1
+    "torch.int32"   : 2, # ncclInt32      = 2
+                         # ncclUint32     = 3
+    "torch.int64"   : 4, # ncclInt64      = 4
+                         # ncclUint64     = 5
+    "torch.float16" : 6, # ncclFloat16    = 6
+    "torch.float32" : 7, # ncclFloat32    = 7
+    "torch.float64" : 8, # ncclFloat64    = 8
+    "torch.bfloat16": 9, # ncclBfloat16   = 9
+}
+
+
+def adaptive_send(tensor, dst, group=None, tag=0):
+    """
+    Sends a tensor together with its `dtype` synchronously.
+
+    Args:
+        tensor (Tensor): Tensor to send.
+        dst (int): Destination rank.
+        group (ProcessGroup, optional): The process group to work on. If None,
+            the default process group will be used.
+        tag (int, optional): Tag to match send with remote recv
+
+    """
+    _check_single_tensor(tensor, "tensor")
+    if _rank_not_in_group(group):
+        _warn_not_in_group("send")
+        return
+
+    send_tensor_size=torch.tensor(tensor.size(),dtype=torch.int).to(tensor.device)
+    send_tensor_dtype=torch.tensor([Tensordtype2ncclDataType_t[str(tensor.dtype)]],dtype=torch.int).to(tensor.device)
+    # send_tensor_info[0:-1] contains the shape of `tensor`
+    # send_tensor_info[-1] is the dtype of `tensor`
+    send_tensor_info=torch.cat((send_tensor_size,send_tensor_dtype))
+    send_len_tensor_info=torch.tensor(send_tensor_info.size(),dtype=torch.int).to(tensor.device)
+
+    if group is None or group is GroupMember.WORLD:
+        default_pg = _get_default_group()
+        default_pg.send([send_len_tensor_info], dst, tag=tag-1).wait()
```

Comment:
Is it safe to use tag-1? What if a previous operation is using tag-1?

Reply:
This is also my main concern. Do you have any suggestion on this?